# OPERAZIONI BASILARI

In [1]:
2 + 3 + 9 # addizzione

14

In [2]:
99 - 73 # sottrazione

26

In [3]:
23.54 * -1432 #moltiplicazione  valori decimali

-33709.28

In [4]:
100 / 7 # divisone

14.285714285714286

In [5]:
100 // 7 # quante volte il 7 sta nel 100?

14

In [6]:
100 % 7 # quanto rimane se divido 100 per il numero di volte massimo che il 7 ci sta pienamente?

2

In [7]:
5 ** 3 # elevazione a potenza

125

Ovviamente, divisioni e moltiplicazioni hanno la precedenza su addizioni e sottrazioni.

In [8]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

# DEFINIZIONE VARIABILI

Salvare delle variabili ci può aiutare a risolvere problemi più complessi.

Si consideri il seguente esempio:

    Un negozio vende delle sacche di ghiaccio a 1.25 dollari e fa il 20% del profitto. Se vende 500 borse di ghiaccio, quanto profitto genera?

In [9]:
500 * (.2 * 1.25)

125.0

Quindi il negozio genera 125 dollari di profitto. Per rendere il problema più "gestibile" conviene memorizzarsi questi numeri in alcune variabili.

In [10]:
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

Calcoliamo quindi il profitto generato e lo printiamo a schermo.

In [11]:
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag
total_profit

125.0

Se chiedo di printare una variabile non definita ottengo un errore.

In [12]:
net_profit

NameError: name 'net_profit' is not defined

Possiamo anche concatenare stringhe e numeri. Per printare a schermo si usa la funzioen 'print()'.

## DEF: Funzione

Una funzione è un set di istruzioni riutilizzabile.
Essa prende uno o più input, performa delle specifiche azioni, e restituisce un output. 
    
Python fornisce molte funzioni in-build come **print** e permette di definire le proprie funzioni.

Possiamo anche concatenare stringhe e numeri.

In [ ]:
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


# CONDIZIONI

In [ ]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [ ]:
my_favorite_number == 1

True

In [ ]:
my_favorite_number == my_least_favorite_number

False

In [ ]:
my_favorite_number != a_neutral_number

True

In [ ]:
my_least_favorite_number > a_neutral_number

True

In [ ]:
my_favorite_number < 10

True

In [ ]:
my_favorite_number >= 1

True

In [ ]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

Is the ice bag expensive? False


# COMBINAZIONI DI CONDIZIONI CON OPERATORI LOGICI

Gli operatori logici `and`, `or` e `not` operano sotto condizioni booleane `True` & `False`.
Nello speicifico `and` e `or` operano due condizioni, mentre `not` opera su una singola condizione.

### Operatore `and`
L'operatore `and` ritorna `True` quando entrambe le condizioni sono `True`. ALtrimenti, ritorna `False`.

In [ ]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [ ]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [ ]:
my_favorite_number > 0 and my_favorite_number >= 3

False

In [ ]:
True and False

False

In [ ]:
True and True

True

### Operatore `or`
L'operatore `or` ritorna `True` se almeno una delle due condizioni è `True`. Ritorna invece `False` se entrambe le condizioni sono `False`.

In [ ]:
a_neutral_number = 3
a_neutral_number == 3 or my_favorite_number < 0

False

In [ ]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [ ]:
my_favorite_number < 0 or True

True

In [ ]:
False or False

False

### Operatore `not`
L'operatore `not` ritorna `False` se la condizione è `True`, mentre ritorna `True` se la condizione è `False`. 

In [ ]:
not a_neutral_number == 3

False

In [ ]:
not a_neutral_number < 0

True

In [ ]:
not False

True

In [ ]:
not True

False

## Combinazione di condizioni

Gli operatori visti precedentemente, possono essere combinati insieme per avere delle condizioni più complesse. Per farlo sfrutteremo le parentesi `(` e `)` per indicare in che ordine verificare le condizioni (ovviamente prima le considizioni nelle parentesi).

In [ ]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

Il risultato è `True` poichè la condizione `or` richiede che una delle due condizioni sia `True`. Si noti che la parte `not` restrituisce proprio un `True`.

In [ ]:
not (True and 0 < 1) or (False and True)

Se non ci sono le parentesi `(` `)`, allora gli operatori logici leggono le condizioni andando da sinistra a destra.

In [ ]:
not True and 0 < 1 or False and True

# SALVARE E CARICARE I MIEI NOTEBOOK (JOVIAN Non funge)

Ci servirà installare un pacchetto specifico: `jovian`.
Per installarlo, runniamo nel terminal di python la stringa `!pip install jovian --upgrade --quiet`. 

In [276]:
!pip install jovian --upgrade --quiet

You should consider upgrading via the 'C:\Users\daniele\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip' command.


E quindi passiamo ad importare questa libreria.

In [13]:
import jovian

<IPython.core.display.Javascript object>

In [15]:
jovian.configure()

[jovian] If you're a jovian-pro user please enter your company's organization ID on Jovian (otherwise leave it blank).
Organization ID:

Abort: 

In [ ]:
eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJmcmVzaCI6ZmFsc2UsImlhdCI6MTYzNDgzMDE0NywianRpIjoiNmZhZWE1OTItZGIzNC00ZTQxLWFkNzYtY2FmNzBmMjJmMDU0IiwidHlwZSI6ImFjY2VzcyIsImlkZW50aXR5Ijp7ImlkIjoxMjU0NjcsInVzZXJuYW1lIjoidGhlc3RhdGlzdGljd2VudGRvd250b2dpdGh1YiJ9LCJuYmYiOjE2MzQ4MzAxNDcsImV4cCI6MTYzODcxODE0N30.mnP3TvHGUAnF09IClp6tQZN6FuiFY7vBo1yGwPFL3R8

Finalmente, possiamo runnare il comando `jovian.commit` per catturarare e uplodare uno snapshot del notebook.

In [272]:
jovian.commit(project="Lesson 1 - Basic operation.ipynb")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

[jovian] Error: Failed to read the Jupyter notebook. Please re-run this cell to try again. If the issue persists, provide the "filename" argument to "jovian.commit" e.g. "jovian.commit(filename='my-notebook.ipynb')"
